### PR TITLE
Fix failing detection of view params bugfix

### PR DESCRIPTION
**Pull request by janciesko**
This is a stab at resolving the issue [View instantiation segfaults when instantiating with wrong number of args](https://github.com/kokkos/kokkos/issues/4561).

This PR moves the runtime_check_rank to the View constructor overload that calls into m_map.allocate_shared and places the check before m_map.allocate_shared but after m_track and m_map allocation. A view parameter is added that allows to turn the check on or off depending on the code path.  

Changes:
- Move runtime_check_rank to the allocating View constructor
- Initializes default dimensions to KOKKOS_IMPL_CTOR_DEFAULT_ARG in Kokkos_Layout (drive-by change)
- Add View param to disable param check for some code paths (see note below)
- PR includes the extra unit-test so we make sure this is doing what it we think it is doing

Note: Disabling runtime_check_rank for some code paths, e.g. TestViewLayoutStrideAssignment.hpp:638 is required and would fail otherwise:
[ RUN      ] cuda_DeathTest.view_layoutstride_right_to_layoutleft_assignment
view_layoutstride_right_to_layoutleft_assignment: srand(1639811970)
Constructor for Kokkos View 'LayoutStride_mirror' has mismatched number of arguments. Number of arguments = 8 but dynamic rank = 1

### CODE REVIEW

**Comment by janciesko** [comment]
Retest this please.

File: core/unit_test/TestViewCtrDimMatch.hpp [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+// ************************************************************************
+//@HEADER
+*/
+
+#include <gtest/gtest.h>
+#include <Kokkos_Core.hpp>
+
+namespace Test {
+
+#define LIVE(EXPR, ARGS, DYNRANK)
```
Comment (dalg24) [inline core/unit_test/TestViewCtrDimMatch.hpp:50]: Why does it resolve to nothing.  Shouldn't we actually construct the view?
I suppose you could do `EXPECT_NO_THROW(EXPR)`

**Comment by dalg24** [comment]
Fix the format

File: core/src/Kokkos_View.hpp [hunk 2]
```diff
@@ -132,6 +132,66 @@ void runtime_check_rank_host(const size_t dyn_rank, const bool is_void_spec,
 }
 #endif
 
+template <typename Traits, typename Layout>
+KOKKOS_INLINE_FUNCTION void runtime_check_rank(
```
Comment (crtrott) [inline core/src/Kokkos_View.hpp:136]: do the same as in the above: instead of templating on Traits: add a parameter for `dyn_rank` and `is_void_spec`. Essentially make the interface like the `runtime_check_rank_host` just that you get a Layout instead of integers.

File: core/src/Kokkos_View.hpp [hunk 3]
```diff
@@ -132,6 +132,66 @@ void runtime_check_rank_host(const size_t dyn_rank, const bool is_void_spec,
...
+  Impl::runtime_check_rank_host(
+      Traits::rank_dynamic,
+      std::is_same<typename Traits::specialize, void>::value, i0, i1, i2, i3,
+      i4, i5, i6, i7, label);
+#endif
+}
+
+template <typename Traits, typename Layout>
+KOKKOS_INLINE_FUNCTION void runtime_check_rank(
+    Layout& layout) {
```
Comment (crtrott) [inline core/src/Kokkos_View.hpp:167]: lets remove this one and just call the other one with an empty label if necessary

File: core/src/Kokkos_View.hpp [hunk 4]
```diff
@@ -132,6 +132,66 @@ void runtime_check_rank_host(const size_t dyn_rank, const bool is_void_spec,
...
+  size_t i1 = layout.dimension[1];
+  size_t i2 = layout.dimension[2];
+  size_t i3 = layout.dimension[3];
+  size_t i4 = layout.dimension[4];
+  size_t i5 = layout.dimension[5];
+  size_t i6 = layout.dimension[6];
+  size_t i7 = layout.dimension[7];
+
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
+  KOKKOS_IMPL_IF_ON_HOST
```
Comment (crtrott) [inline core/src/Kokkos_View.hpp:148]: after the #4660 is in you should use that.

File: core/src/Kokkos_View.hpp [hunk 5]
```diff
@@ -1548,6 +1609,18 @@ class View : public ViewTraits<DataType, Properties...> {
 #endif
     //------------------------------------------------------------
 
+    if (check_input_args) {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
```
Comment (crtrott) [inline core/src/Kokkos_View.hpp:1613]: this is not necessary: the consturctor is never called on the device. Note it is not marked as a KOKKOS_FUNCTION

File: core/src/Kokkos_View.hpp [hunk 6]
```diff
@@ -1497,7 +1557,8 @@ class View : public ViewTraits<DataType, Properties...> {
       const Impl::ViewCtorProp<P...>& arg_prop,
       typename std::enable_if<!Impl::ViewCtorProp<P...>::has_pointer,
                               typename traits::array_layout>::type const&
-          arg_layout)
+          arg_layout,
+      int check_input_args = 0)
```
Comment (crtrott) [inline core/src/Kokkos_View.hpp:1561]: is this ever called in a context where we don't want the check?

File: core/unit_test/TestViewCtrDimMatch.hpp (hunk 1 above)
Comment (crtrott) [reply core/unit_test/TestViewCtrDimMatch.hpp:50]: I like that.

**Review by crtrott — DISMISSED** [review]
See my comments, but basically good.

File: core/src/Kokkos_View.hpp (hunk 6 above)
Comment (janciesko) [reply core/src/Kokkos_View.hpp:1482]: Yes. TestViewLayoutStrideAssignment.hpp:638 would fail when including the check

File: core/src/Kokkos_View.hpp (hunk 4 above)
Comment (janciesko) [reply core/src/Kokkos_View.hpp]: Yes - @dalg24, please rebase #4660 once that PR is ready so we catch this ifdef and update.

File: core/src/Kokkos_View.hpp (hunk 5 above)
Comment (janciesko) [reply core/src/Kokkos_View.hpp]: Removed.

File: core/src/Kokkos_View.hpp (hunk 3 above)
Comment (janciesko) [reply core/src/Kokkos_View.hpp]: Ok, simplified the call graph. Also removed templating.

File: core/unit_test/TestViewCtrDimMatch.hpp (hunk 1 above)
Comment (janciesko) [reply core/unit_test/TestViewCtrDimMatch.hpp]: Done

**Comment by janciesko** [comment]
Comments addressed. Thanks! Once CI passes, we should be good to go.

**Comment by janciesko** [comment]
Retest this please.

**Comment by janciesko** [comment]
Ok, DEATH_TEST is failing for OMPTarget, dynamic params are not passed correctly so we do not catch the incorrect occurrences. This might be an issue in that backend.

File: core/src/Kokkos_View.hpp [hunk 7]
```diff
@@ -79,9 +79,10 @@ class ViewMapping {
 };
 
 template <typename IntType>
-KOKKOS_INLINE_FUNCTION std::size_t count_valid_integers(
-    const IntType i0, const IntType i1, const IntType i2, const IntType i3,
-    const IntType i4, const IntType i5, const IntType i6, const IntType i7) {
+inline std::size_t count_valid_integers(const IntType i0, const IntType i1,
```
[the remark this answers is not in the document]
Comment (janciesko) [reply core/src/Kokkos_View.hpp:82]: We're calling this from a host function only thus the annotation creates a compiler warning.

[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/Kokkos_View.hpp:82]: Calling a `__host__ __device__` function from a `__host__` function should be fine.  I expect the inverse is what would yield a warning.

[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Kokkos_View.hpp:82]: Does that make sense? I would only expect a warning if we call a host-only function from a `host`-`device` function but not if we call a `host-device` function from a host function.

File: core/src/Kokkos_View.hpp [hunk 8]
```diff
@@ -1497,7 +1478,8 @@ class View : public ViewTraits<DataType, Properties...> {
       const Impl::ViewCtorProp<P...>& arg_prop,
       typename std::enable_if<!Impl::ViewCtorProp<P...>::has_pointer,
                               typename traits::array_layout>::type const&
-          arg_layout)
+          arg_layout,
+      int check_input_args = 0)
```
Comment (masterleinad) [inline core/src/Kokkos_View.hpp:1482]:
```suggestion
      bool check_input_args = false)
```

File: core/src/Kokkos_View.hpp [hunk 9]
```diff
@@ -1577,7 +1579,8 @@ class View : public ViewTraits<DataType, Properties...> {
       const Impl::ViewCtorProp<P...>& arg_prop,
       typename std::enable_if<Impl::ViewCtorProp<P...>::has_pointer,
                               typename traits::array_layout>::type const&
-          arg_layout)
+          arg_layout,
+      int check_input_args = 0)
```
Comment (masterleinad) [inline core/src/Kokkos_View.hpp:1583]:
```suggestion
      bool check_input_args = false)
```
Why isn't true the default?

File: core/src/Kokkos_View.hpp (hunk 6 above)
Comment (janciesko) [reply core/src/Kokkos_View.hpp]: I've added an enum, good point Damien. Daniel, we need to disable the check for Views constructions coming from some other funky places like LayoutStride_mirror (TestViewLayoutStrideAssignment.hpp:638).

File: core/src/Kokkos_View.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (janciesko) [reply core/src/Kokkos_View.hpp:82]: My comment was imprecise: We're get the compiler warning from using std::string in a `__host__` `__device__` function.

File: core/unit_test/TestViewCtrDimMatch.hpp [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+#define PARAM_7 1, 1, 1, 1, 1, 1, 1
+
+#define PARAM_0_RANK 0
+#define PARAM_1_RANK 1
+#define PARAM_2_RANK 2
+#define PARAM_3_RANK 3
+#define PARAM_4_RANK 4
+#define PARAM_5_RANK 5
+#define PARAM_6_RANK 6
+#define PARAM_7_RANK 7
```
[the remark this answers is not in the document]
Comment (janciesko) [reply core/unit_test/TestViewCtrDimMatch.hpp]: Fixed

File: core/src/Kokkos_View.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Kokkos_View.hpp:82]: Ah, but that only applies to `runtime_check_rank`, right? Anyway, I think it's fine since that is the only function to call this one anyway.

[the remark this answers is not in the document]
Comment (janciesko) [reply core/src/Kokkos_View.hpp:82]: Yea, thanks Daniel for looking at that. The troublesome issue that I am having right now is that some call path from TestViewLayoutStrideAssignment.hpp:638 is giving `Constructor for Kokkos View 'LayoutStride_mirror' has mismatched number of arguments. Number of arguments = 8 but dynamic rank = 1` **if** we always check the parameters, that is, if we remove that conditional in the allocating View ctor. I am not sure if this is wanted or if we are hiding an issue which also did not exhibit in the previous implementation. There is likely an issue because CI shows errors and also this reported issue: https://github.com/kokkos/kokkos/issues/2995
This seems to be related.
More details to the CI: failing test:
```
[ RUN ] cuda_DeathTest.view_layoutstride_right_to_layoutleft_assignment
view_layoutstride_right_to_layoutleft_assignment: srand(1639811970)
Constructor for Kokkos View 'LayoutStride_mirror' has mismatched number of arguments. Number of arguments = 8 but dynamic rank = 1
```

**Comment by janciesko** [comment]
Don't merge until #2995 is resolved.

**Comment by janciesko** [comment]
Retest this please

**Comment by janciesko** [comment]
Retest this please

**Comment by janciesko** [comment]
Retest this please

File: core/unit_test/TestViewCtorDimMatch.hpp [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+#define PARAM_7_RANK 7
+
+using DType = int;
+
+// Skip test execution when KOKKOS_ENABLE_OPENMPTARGET is enabled until
+// Kokkos::abort() aborts properly on that backend
+#if defined(KOKKOS_ENABLE_OPENMPTARGET)
+#else
+TEST(TEST_CATEGORY_DEATH, view_construction_with_wrong_params_dyn) {
+  GTEST_FLAG_SET(death_test_style, "threadsafe");
```
Comment (cz4rs) [inline core/unit_test/TestViewCtorDimMatch.hpp:83]:
```
/var/jenkins/workspace/Kokkos/core/unit_test/TestViewCtorDimMatch.hpp: In member function 'virtual void Test::openmp_DeathTest_view_construction_with_wrong_params_dyn_Test::TestBody()':

/var/jenkins/workspace/Kokkos/core/unit_test/TestViewCtorDimMatch.hpp:83:18: error: 'death_test_style' was not declared in this scope

   GTEST_FLAG_SET(death_test_style, "threadsafe");
```

Comment (cz4rs) [reply core/unit_test/TestViewCtorDimMatch.hpp:83]: As far as I can see we are missing `GTEST_FLAG_SET` completely in our copy of gtest.
Anyways, looking at the implementation the previous version should work just fine:
https://github.com/google/googletest/blob/25ad42aabeaba6227f37795cdd2752e128e83827/googletest/include/gtest/internal/gtest-port.h#L2219

Comment (janciesko) [reply core/unit_test/TestViewCtorDimMatch.hpp:83]: This should work. I'll amend this.
`GTEST_FLAG_SET(GTEST_DEFAULT_DEATH_TEST_STYLE, "threadsafe"); `

Comment (janciesko) [reply core/unit_test/TestViewCtorDimMatch.hpp]: I've rolled back the commit.

**Comment by cz4rs** [comment]
No idea what exactly could be wrong with this :disappointed: 
To at least get a stacktrace, you could try replacing `ctest --verbose` at the end of `CUDA-11.4-NVHPC` build:
https://github.com/kokkos/kokkos/blob/07809a79e93a67d38673b078719f05824412de6b/.jenkins#L54
with:
```
catchsegv /var/jenkins/workspace/Kokkos/build/core/unit_test/KokkosCore_UnitTest_OpenMP
```
(as described in [this SO answer](https://stackoverflow.com/a/6599348)).

`catchsegv` is available by default in Ubuntu, so maybe this can help somehow.

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

**Comment by janciesko** [comment]
@cz4rs, all. I was able to dig into this. I have documented the issue here: https://github.com/google/googletest/issues/3760
Seems to be specific to NVHPC and when handling std::abort (or std::terminate()) within the root process.

**Comment by janciesko** [comment]
Skipping the unit test execution when KOKKOS_COMPILER_NVHPC until we find out more why gtest is segfaulting with nvhpc.

File: core/unit_test/TestViewCtorDimMatch.hpp [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+#define PARAM_5_RANK 5
+#define PARAM_6_RANK 6
+#define PARAM_7_RANK 7
+
+using DType = int;
+
+// Skip test execution when KOKKOS_ENABLE_OPENMPTARGET is enabled until
+// Kokkos::abort() aborts properly on that backend
+// Skip test execution when KOKKOS_COMPILER_NVHPC until fixed in GTEST
+#if defined(KOKKOS_ENABLE_OPENMPTARGET) || (KOKKOS_COMPILER_NVHPC)
```
Comment (dalg24) [inline core/unit_test/TestViewCtorDimMatch.hpp:81]:
```suggestion
#if defined(KOKKOS_ENABLE_OPENMPTARGET) || (KOKKOS_COMPILER_NVHPC)
```

File: core/unit_test/TestTeamVector.hpp [hunk 13]
```diff
@@ -1031,8 +1031,10 @@ TEST(TEST_CATEGORY, parallel_scan_with_reducers) {
   using T = double;
   using namespace VectorScanReducer;
 
-  static constexpr int n              = 1000000;
-  static constexpr int n_vector_range = 100;
+  constexpr int n              = 1000000;
+  constexpr int n_vector_range = 100;
+  (void)n;
+  (void)n_vector_range;
```
Comment (dalg24) [inline core/unit_test/TestTeamVector.hpp:1037]: Why did you change this?

File: core/unit_test/TestTeamVector.hpp [hunk 14]
```diff
@@ -1033,6 +1033,8 @@ TEST(TEST_CATEGORY, parallel_scan_with_reducers) {
 
   static constexpr int n              = 1000000;
   static constexpr int n_vector_range = 100;
+  (void)n;
+  (void)n_vector_range;
```
Comment (dalg24) [inline core/unit_test/TestTeamVector.hpp:1037]:
```suggestion
```

File: core/src/Kokkos_View.hpp [hunk 15]
```diff
@@ -1533,6 +1520,26 @@ class View : public ViewTraits<DataType, Properties...> {
...
+      size_t i7 = arg_layout.dimension[7];
+
+      const std::string& alloc_name =
+          static_cast<Kokkos::Impl::ViewCtorProp<void, std::string> const&>(
+              prop_copy)
+              .value;
+      Impl::runtime_check_rank(
+          traits::rank_dynamic,
+          std::is_same<typename traits::specialize, void>::value, i0, i1, i2,
+          i3, i4, i5, i6, i7, alloc_name.c_str());
```
Comment (dalg24) [inline core/src/Kokkos_View.hpp:1540]:
```suggestion
          i3, i4, i5, i6, i7, alloc_name);
```
Why converting to a C-string if the functions takes `std::string const&`?

File: core/unit_test/TestTeamVector.hpp (hunk 13 above)
Comment (janciesko) [reply core/unit_test/TestTeamVector.hpp]:
```
"/ascldap/users/jciesko/Kokkos/kokkos_viewctr_fix/kokkos3/core/unit_test/TestTeamVector.hpp", line 1034: error: variable "n" was declared but never referenced
  static constexpr int n = 1000000;
                       ^

"/ascldap/users/jciesko/Kokkos/kokkos_viewctr_fix/kokkos3/core/unit_test/TestTeamVector.hpp", line 1035: error: variable "n_vector_range" was declared but never referenced
  static constexpr int n_vector_range = 100;
```

**Comment by janciesko** [comment]
Retest this please

File: core/src/Kokkos_Layout.hpp [hunk 16]
```diff
@@ -84,9 +84,14 @@ struct LayoutLeft {
...
-                                size_t N3 = 0, size_t N4 = 0, size_t N5 = 0,
-                                size_t N6 = 0, size_t N7 = 0)
+  explicit constexpr LayoutLeft(size_t N0 = KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+                                size_t N1 = KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+                                size_t N2 = KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+                                size_t N3 = KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+                                size_t N4 = KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+                                size_t N5 = KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+                                size_t N6 = KOKKOS_IMPL_CTOR_DEFAULT_ARG,
+                                size_t N7 = KOKKOS_IMPL_CTOR_DEFAULT_ARG)
```
Comment (PhilMiller) [inline core/src/Kokkos_Layout.hpp:94]: This may actually address part of the issue I saw here:
https://github.com/kokkos/kokkos/pull/4775

**Review by PhilMiller — APPROVED** [review]

**Review by dalg24 — COMMENTED** [review]
I am worried that we disable validation on the device-side

File: core/src/Kokkos_View.hpp [hunk 17]
```diff
@@ -1577,7 +1584,8 @@ class View : public ViewTraits<DataType, Properties...> {
       const Impl::ViewCtorProp<P...>& arg_prop,
       typename std::enable_if<Impl::ViewCtorProp<P...>::has_pointer,
                               typename traits::array_layout>::type const&
-          arg_layout)
+          arg_layout,
+      check_input_args check_args = check_input_args::no)
```
[the remark this answers is not in the document]
Comment (janciesko) [reply core/src/Kokkos_View.hpp:1588]: ViewCtorProp on device has "has_label" false. Adding the comment, thanks

**Comment by janciesko** [comment]
Retest this please

**Review by cz4rs — APPROVED** [review]

**Comment by cz4rs** [comment]
Note: https://github.com/google/googletest/issues/3760 was dropped with:
> This platform is unsupported.

**Comment by janciesko** [comment]
Retest this please

**Comment by janciesko** [comment]
Retest this please

**Comment by janciesko** [comment]
Retest this please

File: core/src/Kokkos_View.hpp [hunk 18]
```diff
@@ -1403,14 +1390,19 @@ class View : public ViewTraits<DataType, Properties...> {
...
   //----------------------------------------
   // Allocation according to allocation properties and array layout
 
   template <class... P>
   explicit inline View(
       const Impl::ViewCtorProp<P...>& arg_prop,
       std::enable_if_t<!Impl::ViewCtorProp<P...>::has_pointer,
-                       typename traits::array_layout> const& arg_layout)
+                       typename traits::array_layout> const& arg_layout,
+      check_input_args check_args = check_input_args::no)
```
[the remark this answers is not in the document]
Comment (janciesko) [reply core/src/Kokkos_View.hpp:1405]: The idea here is that we catch discrepancies originating from user error. Other code paths exist, such as 
```
create_mirror_view(const Space&, const Kokkos::View<T, P...>& src,
                   const I&... arg_prop) {
  return typename Impl::MirrorViewType<Space, T, P...>::view_type(
      Kokkos::view_alloc(src.label(), arg_prop...), src.layout());
```
where we instatiate a view of a given type (for example int **, LayoutLeft) and where in that case  the src.layout has a layout of {20, 10, 1, 1, 1, 1, 1}. In that case, the check would fail with "Number of arguments = 8 but dynamic rank = 2". So I think this is ok.

File: core/src/Kokkos_View.hpp [hunk 19]
```diff
@@ -1487,8 +1499,10 @@ class View : public ViewTraits<DataType, Properties...> {
   explicit KOKKOS_INLINE_FUNCTION View(
       const Impl::ViewCtorProp<P...>& arg_prop,
       std::enable_if_t<Impl::ViewCtorProp<P...>::has_pointer,
-                       typename traits::array_layout> const& arg_layout)
-      : m_track()  // No memory tracking
+                       typename traits::array_layout> const& arg_layout,
+      check_input_args /*ignored*/ =
+          check_input_args::no)  // not checking on device
```
[the remark this answers is not in the document]
Comment (janciesko) [reply core/src/Kokkos_View.hpp:1504]: I think the comment is unfortunate. It should say, not checking for this View ctr. Similar reason as below.

**Review by masterleinad — APPROVED** [review]
Looks OK to me. This preserves behavior and we could think about extending the checks in a follow-up.

**Review by crtrott — APPROVED** [review]

**Comment by crtrott** [comment]
Follow up: make a test work on device to check for proper error.
